File: collect_app/src/main/java/org/odk/collect/android/formmanagement/matchexactly/SyncStatusAppState.java
package org.odk.collect.android.formmanagement.matchexactly;

import android.content.Context;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import org.odk.collect.android.provider.FormsProviderAPI;
import org.odk.collect.forms.FormSourceException;

import javax.inject.Singleton;

@Singleton
public class SyncStatusAppState {

    private final MutableLiveData<Boolean> syncing = new MutableLiveData<>(false);
    private final MutableLiveData<FormSourceException> lastSyncFailure = new MutableLiveData<>(null);
    private final Context context;

    public SyncStatusAppState(Context context) {
        this.context = context;
    }

    public LiveData<Boolean> isSyncing() {
        return syncing;
    }

    public LiveData<FormSourceException> getSyncError() {
        return lastSyncFailure;
    }

    public void startSync() {
        syncing.postValue(true);
    }

    public void finishSync(@Nullable FormSourceException exception) {
        lastSyncFailure.postValue(exception);
        syncing.postValue(false);
        context.getContentResolver().notifyChange(FormsProviderAPI.CONTENT_URI, null);
    }
}

File: collect_app/src/main/java/org/odk/collect/android/formmanagement/matchexactly/SyncStatusAppState.kt
package org.odk.collect.android.formmanagement.matchexactly

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import org.odk.collect.android.external.FormsContract
import org.odk.collect.androidshared.data.AppState
import org.odk.collect.forms.FormSourceException

class SyncStatusAppState(private val appState: AppState, private val context: Context) {

    fun isSyncing(projectId: String): LiveData<Boolean> {
        return getSyncingLiveData(projectId)
    }

    fun getSyncError(projectId: String): LiveData<FormSourceException?> {
        return getSyncErrorLiveData(projectId)
    }

    fun startSync(projectId: String) {
        getSyncingLiveData(projectId).postValue(true)
    }

    fun finishSync(projectId: String, exception: FormSourceException?) {
        getSyncErrorLiveData(projectId).postValue(exception)
        getSyncingLiveData(projectId).postValue(false)
        context.contentResolver.notifyChange(FormsContract.getUri(projectId), null)
    }

    private fun getSyncingLiveData(projectId: String) =
        appState.get("$KEY_PREFIX_SYNCING:$projectId", MutableLiveData(false))

    private fun getSyncErrorLiveData(projectId: String) =
        appState.get("$KEY_PREFIX_ERROR:$projectId", MutableLiveData<FormSourceException>(null))

    fun clear(projectId: String) {
        getSyncingLiveData(projectId).value = false
        getSyncErrorLiveData(projectId).value = null
    }

    companion object {
        const val KEY_PREFIX_SYNCING = "syncStatusSyncing"
        const val KEY_PREFIX_ERROR = "syncStatusError"
    }
}
